### DISCUSSION TITLE
How Can I Position Something Absolutely in the Page?

### DISCUSSION THREAD

**Discussion by zicklag** [typst/typst#1461] [Q&A]
Hey there! I'm trying out Typst and really loving it so far! I'm so glad that somebody finally made a Latex replacement in Rust. ❤

Anyway, I'm trying to position a line and a circle to follow alongside my content, like a timeline. I'm using the `locate` function to find out where my content is in the page, but I'm getting stuck because I can't put the line at the position returned from `locate`, because `locate` returns relative to the page, and `line`, in combination with `place`, can only be placed relative to it's parent!

Is there any way around this?

**Comment by PgBiel** [comment]
More context might be needed here, but, in general, you should be able to call `locate` on the parent as well and subtract their coordinates to have a similar effect. (Ensure you're using `place(top+left, dx: ..., dy: ...)` in that case, since absolute coordinates are relative to the top left of the page.)

**Reply by zicklag** [reply]
Oh, yeah, I could do that. If you had a lot of nesting I could imagine that being rather cumbersome, but it should work. I'll try it, thanks!

**Comment by sivsivsree** [comment]
How to get the coordinates of the placed document?

**Comment by aaeberharter** [comment]
I am having a similar issue and thus I am posting in this thread to make the question more concrete.

The function `leftnote` is supposed to be called anywhere in the document and add the given content beyond the left margin, what I have so far is this:
```
#let leftnote(body) = context {
  place(top + left, dx:-2cm, dy:here().position().y, body)
}
```

This code snippet has 2 issues:

1. `here().position()` gives us absolute page coordinates, however `place` takes relative coordinates to its parent. It is unclear to me how I could get the position of the parent inside `leftnote` in order to perform the required math
2. Calling the function like `Hel#leftnote[note]lo` adds an undesired linebreak, in general the surrounding content is disturbed, see image.
![image](https://github.com/user-attachments/assets/b6c414d9-e80c-461f-a70c-d8b6cc05e26b)

**Reply by aaeberharter** [reply]
by trial and error I was able to achieve significant progress:

```
#let leftnote(body) = context {
  let outer_pos = here().position()             // store the position in the caller context
  box[#context place(                           // create a box and switch into its context
    dx:here().position().x - outer_pos.x,       // now the magic happens: we call #place and put as displacement the
    dy:here().position().y - outer_pos.y        //    difference of the current context and the stored caller context.
    )[                                          //    This magically places the content at the page origin (0,0).
      #place(dx: 0.5cm, dy: outer_pos.y)[#body]]// dx and dy now take absolute values in the page coordinate system
  ]
}
```
I would be very thankful towards anyone who can explain the magic to me, why does it move "the cursor" to the page origin regardless of the element depth of the caller?

One remaining problem is the exact positioning of the content, consider the following call to `leftnote`:

```
#leftnote[Hello]
World

#leftnote[Hello]World
```
![image](https://github.com/user-attachments/assets/8ac134c3-7a11-43bf-8d4d-05501689b65f)


1. In the first case the linebreak in typst source causes an intendation of the "World", this does not happen if function call and content are on the same line.
2. In both cases the top of "Hello" is flush with the bottom of the caller context, I would like to have it flush with the top.

**Comment by hurzl** [comment]
I'm not sure if I understand the problem correctly.
This places the note 1 cm from left page edge in a 3 cm wide box:
```
#let leftnote(body) = context {
  let outer_pos = here().position()   // store the position in the caller context
  box[#place(dx:1cm-outer_pos.x,dy:-0.65em)[#box(width:3cm)[#body]]]
}
```

**Reply by aaeberharter** [reply]
You are completely right, I confused myself with all the `context` tags.

I am now using the following code
```
#let links(body) = context {
  let outer_pos = here().position()
  let y_correction = 3.0 * (par.leading - text.size)
  box(place(dx:1cm-outer_pos.x,dy:y_correction, box(width:2.5cm, align(right,body)))) + h(0pt, weak: true)
}
```
The trailing zero space with parameter `weak` prevents undesired indentations.

The post of @hurzl should be markes as the answer.
